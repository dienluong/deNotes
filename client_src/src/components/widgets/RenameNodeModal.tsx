import React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';

// Types
import { FormEvent, ChangeEvent } from 'react';
export type PropsT = {
  nodeType: NodeTypeT,
  currentName: string,
  onCloseHandler: () => unknown;
  onSubmitHandler: ({ name }: { name: string }) => unknown;
};

export default function RenameNodeModal({ currentName, onCloseHandler, onSubmitHandler }: PropsT) {
  const [open, setOpen] = React.useState(true);
  let textFieldValue = currentName;

  function handleClose() {
    setOpen(false);
    if (typeof onCloseHandler === 'function') {
      onCloseHandler();
    }
  }

  function handleChange(event: ChangeEvent<HTMLInputElement> ) {
    textFieldValue = event.target.value;
  }

  function submit(event: FormEvent<HTMLElement>) {
    handleClose();
    // TODO: Trim whitespaces
    if (typeof onSubmitHandler === 'function') {
      onSubmitHandler({ name: textFieldValue });
    }
    event.preventDefault();
  }

  return (
    <div>
      <Dialog
        open={ open }
        onClose={ handleClose }
        aria-labelledby="rename-node-dialog-title"
      >
        <DialogTitle id="rename-node-dialog-title">{ `Enter Name for ${currentName}` }</DialogTitle>
        <DialogContent>
          <form onSubmit={ submit } noValidate autoComplete="off">
            <TextField
              autoFocus
              margin="dense"
              id="name"
              label="Name"
              type="text"
              required
              fullWidth
              variant="outlined"
              defaultValue={ currentName }
              onChange={ handleChange }
              onFocus={ (e: React.FocusEvent<HTMLInputElement>) => e.target.select() }
            />
          </form>
        </DialogContent>
        <DialogActions>
          <Button aria-label={ 'Cancel Rename Dialog' } onClick={ handleClose } color="primary">
            Cancel
          </Button>
          <Button aria-label={ 'OK Rename Dialog' } onClick={ submit } color="primary">
            OK
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}
