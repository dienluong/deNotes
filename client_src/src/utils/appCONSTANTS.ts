interface nodeTypesT {
  FOLDER: Readonly<string>;
  ITEM: Readonly<string>;
}
export const nodeTypes: Readonly<nodeTypesT> = {
  FOLDER: 'FOLDER_NODE',
  ITEM: 'ITEM_NODE',
};

export const NONE_SELECTED: Readonly<string> = '<NO_NODE_SELECTED>';
