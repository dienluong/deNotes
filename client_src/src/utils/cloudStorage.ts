import Widget from 'remotestorage-widget';
import RemoteStorage from 'remotestoragejs';

// Types
type RemoteStorageT = typeof RemoteStorage;

const options = {
  leaveOpen: false,
  autoCloseAfter: 1500,
  skipInitial: true,
  logging: false,
};

export function connect({ storage }: { storage: RemoteStorageT  }) {
  if (storage) {
    const widget = new Widget(storage, options);
    storage.setApiKeys({
      dropbox: process.env.REACT_APP_DROPBOX_APPKEY,
    });
    return {
      storage,
      widget,
    }
  } else {
    return null;
  }
}
