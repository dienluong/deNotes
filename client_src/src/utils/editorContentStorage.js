let _save = () => Promise.reject(new Error('Save aborted. Cause: No Storage implementation provided.'));
let _load = () => Promise.reject(new Error('Load aborted. Cause: No Storage implementation provided.'));
let _remove = () => Promise.reject(new Error('Delete aborted. Cause: No Storage implementation provided.'));

export function inject({ save = _save, load = _load, remove = _remove }) {
  _save = typeof save === 'function' ? save : _save;
  _load = typeof load === 'function' ? load : _load;
  _remove = typeof remove === 'function' ? remove : _remove;
}

export function save({ userId, editorContent }) {
  if (!editorContent || !editorContent.content || !editorContent.delta || !editorContent.id) {
    return Promise.reject(new Error('Save aborted. Cause: invalid content.'));
  }

  if (!userId) {
    return Promise.reject(new Error('Save aborted. Cause: invalid userId.'));
  }

  return _save({
    collectionName: 'notes',
    id: editorContent.id,
    ownerId: userId,
    // TODO: Replace hardcoded values
    dataObj: {
      'id': editorContent.id,
      'title': editorContent.title || '',
      'body': editorContent.content,
      'delta': editorContent.delta,
      'dateCreated': editorContent.dateCreated,
      'dateModified': editorContent.dateModified,
      'ownerId': userId,
    },
  });
}

/**
 * Returns a promise for an array of content
 * @param noteId
 * @return {Promise<Response | never>}
 */
export function load({ id = '', userId = '' }) {
  if (!id && !userId) {
    return Promise.reject(new Error('Load aborted. Cause: invalid parameters.'));
  }

  return _load({
    collectionName: 'notes',
    id,
    ownerId: userId,
  }).then(content => {
    const propList = ['id', 'title', 'body', 'delta', 'dateCreated', 'dateModified'];
    if (content && (typeof content.hasOwnProperty === 'function') && propList.every(prop => content.hasOwnProperty(prop))) {
      // Note: We convert the retreived dates to milliseconds since Unix epoch.
      return {
        id,
        title: content.title,
        content: content.body,
        delta: typeof content.delta === 'string' ? JSON.parse(content.delta) : content.delta,
        dateModified: new Date(content.dateModified).getTime(),
        dateCreated: new Date(content.dateCreated).getTime(),
      };
    } else {
      const message = `Unrecognized data fetched. ID: ${id}`;
      return Promise.reject(new Error(message));
    }
  });
}

export function remove({ ids = '', userId = '' }) {
  if (typeof ids === 'string' || (Array.isArray(ids) && ids.length)) {
    return _remove({
      collectionName: 'notes',
      ids,
      ownerId: userId,
    });
  } else {
    return Promise.reject(new Error('Delete aborted. Cause: invalid parameter.'));
  }
}

