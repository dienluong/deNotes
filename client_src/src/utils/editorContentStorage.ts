let _save: Function = () => Promise.reject(new Error('Save aborted. Cause: No Storage implementation provided.'));
let _load: Function = () => Promise.reject(new Error('Load aborted. Cause: No Storage implementation provided.'));
let _remove: Function = () => Promise.reject(new Error('Delete aborted. Cause: No Storage implementation provided.'));

export function inject({ save = _save, load = _load, remove = _remove }: { save: Function, load: Function, remove: Function })
  : void {
  _save = typeof save === 'function' ? save : _save;
  _load = typeof load === 'function' ? load : _load;
  _remove = typeof remove === 'function' ? remove : _remove;
}

export function save({ userId, editorContent }: { userId: string, editorContent: EditorContentT })
  : Promise<any> {
  if (!editorContent || !editorContent.content || !editorContent.delta || !editorContent.id) {
    return Promise.reject(new Error('invalid content.'));
  }

  if (!userId || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid userId.'));
  }

  return _save({
    collectionName: 'notes',
    id: editorContent.id,
    ownerId: userId,
    dataObj: {
      'id': editorContent.id,
      'title': editorContent.title || '',
      'body': editorContent.content,
      'delta': editorContent.delta,
      'dateCreated': editorContent.dateCreated || 0,
      'dateModified': editorContent.dateModified || 0,
      'ownerId': userId,
    },
  });
}

/**
 * Returns a promise for an array of content
 * @param noteId
 * @return {Promise<Response | never>}
 */
export function load({ id = '', userId = '' }: { id: string, userId: string })
  : Promise<EditorContentT> {
  if (!id || !userId || typeof id !== 'string' || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid parameters.'));
  }

  return _load({
    collectionName: 'notes',
    id,
    ownerId: userId,
  }).then((content: any) => {
    const propList = ['id', 'title', 'body', 'delta', 'dateCreated', 'dateModified'];
    if (content && (typeof content.hasOwnProperty === 'function') && propList.every(prop => content.hasOwnProperty(prop))) {
      // Note: We convert the retreived dates to milliseconds since Unix epoch.
      return {
        id,
        title: content.title,
        content: content.body,
        delta: typeof content.delta === 'string' ? JSON.parse(content.delta) : content.delta,
        dateModified: new Date(content.dateModified).getTime(),
        dateCreated: new Date(content.dateCreated).getTime(),
      };
    } else {
      const message = `Unrecognized data fetched. ID: ${id}`;
      return Promise.reject(new Error(message));
    }
  });
}

export function remove({ ids = '', userId = '' }: { ids: string, userId: string })
  : Promise<{ results: string|string[], count: number }> {
  if (!ids || !userId || (typeof ids !== 'string' && !Array.isArray(ids)) || (Array.isArray(ids) && !ids.length) || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid parameters.'));
  } else {
    return _remove({
      collectionName: 'notes',
      ids,
      ownerId: userId,
    });
  }
}

