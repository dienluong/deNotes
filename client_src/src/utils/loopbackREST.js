/**
 * @param id
 * @param dataObj
 * @return {Promise<Response | never>}
 */
export function save(id, dataObj) {
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/trees/upsertWithWhere?where={"ownerId": ${id}}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(dataObj),
    }
  );
}

/**
 * @param id
 * @return {Promise<Response | never>}
 */
export function load(id) {
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/trees?where={"ownerId": ${id}}`,
    {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    },
  );
}
