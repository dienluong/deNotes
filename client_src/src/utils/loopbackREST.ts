/** Update matching record or create new one if none exists. If both id and ownerId provided, ownerId is ignored.
 * @param collectionName {string} Name of the table or collection
 * @param id {string}
 * @param ownerId {string}
 * @param dataObj
 * @return {Promise<Response | never>}
 */
export function save({ collectionName = '', id = '', ownerId = '', dataObj }
  : { collectionName: string, id: string, ownerId: string, dataObj: object })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Save aborted: invalid collection name.'));
  }
  if (!id && !ownerId) {
    return Promise.reject(new Error('Save aborted: no valid ID provided.'));
  }

  let filter = '';
  if (id) {
    filter = '?where=' + encodeURIComponent(`{"id":"${id}"}`);
  } else if (ownerId) {
    filter = '?where=' + encodeURIComponent(`{"ownerId":"${ownerId}"}`);
  }

  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}/upsertWithWhere${filter}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(dataObj),
    }
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR saving data to remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

/**
 * If multiple items match, results are returned in an array.
 * @param collectionName {string} Name of the table or collection
 * @param id {string}
 * @param ownerId {string}
 * @return {Promise<Response | never>}
 */
export function load({ collectionName = '', id = '', ownerId = '' }: { collectionName: string, id: string, ownerId: string })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Load aborted: invalid collection name.'));
  }
  const ownerFilter = ownerId ? '?where=' + encodeURIComponent(`{"ownerId": "${ownerId}"}`) : '';
  const idFilter = id ? `/${id}` : '';
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}${idFilter}${ownerFilter}`,
    {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
      },
    },
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR loading data from remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

export function remove({ collectionName = '', ids = '' }: { collectionName: string, ids: string|string[] })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Delete aborted: invalid collection name.'));
  }

  let idFilter;
  if (ids && typeof ids === 'string') {
    idFilter = `/${ids}`;
  } else if (Array.isArray(ids) && ids.length) {
    idFilter = `/multidelete?ids=${encodeURIComponent(JSON.stringify(ids))}`;
  } else {
    return Promise.reject(new Error('Delete aborted: invalid filter'));
  }

  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}${idFilter}`,
    {
      method: 'DELETE',
      headers: {
        'Accept': 'application/json',
      },
    },
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR deleting item from remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

