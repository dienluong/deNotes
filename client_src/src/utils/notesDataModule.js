const moduleName = 'notes';

function builder(privateClient) {
  privateClient.declareType(moduleName, {
    'type': 'object',
    'properties': {
      'id': {
        'type': 'string',
        'pattern': '^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$',
      },
      'title': {
        'type': 'string',
      },
      'content': {
        'type': 'string',
      },
      'delta': {
        'type': 'object',
        'properties': {
          'ops': {
            'type': 'array',
            'items': {
              'type': 'object',
            },
          },
        },
      },
      'dateCreated': {
        'type': 'integer',
      },
      'dateModified': {
        'type': 'integer',
      },
      'readOnly': {
        'type': 'boolean',
      },
    },
    'required': ['id'],
  });

  const notesDecorator = {
    save({ ownerId, id, dataObj }) {
      const path = `${ownerId}/${id}`;
      return this.storeObject(moduleName, path, dataObj);
    },
    load({ ownerId, id }) {
      const path = `${ownerId}/${id}`;
      return this.getObject(path);
    },
    destroy({ ownerId, ids }) {
      if (Array.isArray(ids)) {
        if (ids.length) {
          // catch any error from remove() to prevent fast-fail by Promise.all()
          return Promise.all(ids.map(id => this.remove(`${ownerId}/${id}`).catch(err => err)));
        } else {
          return Promise.reject(new Error('Invalid ID'));
        }
      } else {
        return this.remove(`${ownerId}/${ids}`);
      }
    },
  };

  const module = {
    privateContent: () => Object.assign(privateClient.scope(`${moduleName}/`), notesDecorator),
  };

  return {
    exports: module,
  };
}


const notesDataModule = {
  name: moduleName,
  builder: builder,
};

export default notesDataModule;
