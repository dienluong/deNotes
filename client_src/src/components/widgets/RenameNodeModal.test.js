import React from 'react';
import RenameNodeModal from './RenameNodeModal';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import { nodeTypes } from '../../utils/appCONSTANTS';
import { createSerializer } from 'enzyme-to-json';
import { render, cleanup, fireEvent } from 'react-testing-library';
// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

// Configure snapshot serializer
expect.addSnapshotSerializer(createSerializer());

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }

  cleanup();
});

it('renders MUI Dialog and TextField containing current name', () => {
  const props = {
    nodeType: nodeTypes.ITEM,
    currentName: 'current test node name',
    onCloseHandler: jest.fn(),
    onSubmitHandler: jest.fn(),
  };

  wrapper = shallow(<RenameNodeModal { ...props } />);
  expect(wrapper).toMatchSnapshot();

  expect(wrapper.find(Dialog).exists()).toBe(true);
  expect(wrapper.find(Dialog).prop('open')).toBe(true);
  expect(wrapper.find(TextField).exists()).toBe(true);
  expect(wrapper.find(TextField).props()).toMatchObject({ defaultValue: props.currentName });
});

it('calls provided close and submit handlers', () => {
  const props = {
    nodeType: nodeTypes.ITEM,
    currentName: 'current test node name',
    onCloseHandler: jest.fn(),
    onSubmitHandler: jest.fn(),
  };

  wrapper = shallow(<RenameNodeModal { ...props } />);

  // Test onClose event
  const dialogWrapper = wrapper.find(Dialog);
  expect(dialogWrapper.exists()).toBe(true);
  const onCloseHandler = dialogWrapper.prop('onClose');
  expect(onCloseHandler).toBeInstanceOf(Function);
  onCloseHandler();
  expect(props.onCloseHandler).toBeCalledTimes(1);
  props.onCloseHandler.mockClear();

  // Test cancel button
  const cancelWrapper = wrapper.find({ 'aria-label': 'Cancel Rename Dialog' });
  expect(cancelWrapper.exists()).toBe(true);
  cancelWrapper.simulate('click');
  expect(props.onCloseHandler).toBeCalledTimes(1);
  props.onCloseHandler.mockClear();

  // Test form submit
  const formWrapper = wrapper.find('form');
  expect(formWrapper.exists()).toBe(true);
  const evt = {
    preventDefault: jest.fn(),
  };
  formWrapper.simulate('submit', evt);
  expect(props.onSubmitHandler).toBeCalledTimes(1);
  expect(props.onSubmitHandler).lastCalledWith({ name: props.currentName });
  // Submit event also closes the Dialog modal
  expect(props.onCloseHandler).toBeCalledTimes(1);
  props.onCloseHandler.mockClear();
  props.onSubmitHandler.mockClear();

  // Test OK button
  const okWrapper = wrapper.find({ 'aria-label': 'OK Rename Dialog' });
  expect(okWrapper.exists()).toBe(true);
  okWrapper.simulate('click', evt);
  expect(props.onSubmitHandler).toBeCalledTimes(1);
  expect(props.onSubmitHandler).lastCalledWith({ name: props.currentName });
  // Submit event also closes the Dialog modal
  expect(props.onCloseHandler).toBeCalledTimes(1);
});

it('returns the newly entered name on submit', () => {
  const props = {
    nodeType: nodeTypes.ITEM,
    currentName: 'current test node name',
    onCloseHandler: jest.fn(),
    onSubmitHandler: jest.fn(),
  };

  const newName = 'this is a new name';

  const { getByValue, getByText } = render(<RenameNodeModal { ...props } />);
  const textInput = getByValue(props.currentName);
  const okButton = getByText('OK');
  fireEvent.change(textInput, { target: { value: newName } });
  fireEvent.click(okButton);
  expect(props.onSubmitHandler).lastCalledWith({ name: newName });
});
