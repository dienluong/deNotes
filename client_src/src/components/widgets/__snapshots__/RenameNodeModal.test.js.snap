// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MUI Dialog and TextField containing current name 1`] = `
<div>
  <WithStyles(Dialog)
    aria-labelledby="rename-node-dialog-title"
    onClose={[Function]}
    open={true}
  >
    <WithStyles(DialogTitle)
      id="rename-node-dialog-title"
    >
      Enter Name for current test node name
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)>
      <form
        autoComplete="off"
        noValidate={true}
        onSubmit={[Function]}
      >
        <TextField
          autoFocus={true}
          defaultValue="current test node name"
          fullWidth={true}
          id="name"
          label="Name"
          margin="dense"
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          select={false}
          type="text"
          variant="outlined"
        />
      </form>
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        aria-label="Cancel Rename Dialog"
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <WithStyles(Button)
        aria-label="OK Rename Dialog"
        color="primary"
        onClick={[Function]}
      >
        OK
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</div>
`;
